File: src/sandbox/shim.py
import os
import sys
import socket
import json
import base64


def main():
    socket_path = os.environ.get("BRIDGE_SOCK", "./workspace/bridge.sock")
    command = os.environ.get("SHIM_COMMAND", "gh")
    args = sys.argv[1:]
    cwd = os.getcwd()

    req = {
        "command": command,
        "args": args,
        "cwd": cwd,
        "env": {"GH_TOKEN": os.environ.get("GH_TOKEN", "")},
    }

    try:
        client = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        client.connect(socket_path)
        client.sendall(json.dumps(req).encode("utf-8"))

        # Read response
        buffer = ""
        while True:
            data = client.recv(4096)
            if not data:
                break
            buffer += data.decode("utf-8")

            while "\n" in buffer:
                line, buffer = buffer.split("\n", 1)
                if not line.strip():
                    continue
                try:
                    msg = json.loads(line)
                    if msg["type"] == "stdout":
                        sys.stdout.buffer.write(base64.b64decode(msg["data"]))
                        sys.stdout.buffer.flush()
                    elif msg["type"] == "stderr":
                        sys.stderr.buffer.write(base64.b64decode(msg["data"]))
                        sys.stderr.buffer.flush()
                    elif msg["type"] == "exit":
                        sys.exit(msg["code"])
                    elif msg["type"] == "error":
                        print(f"[Shim Error] {msg['message']}", file=sys.stderr)
                        sys.exit(1)
                except Exception as e:
                    # Partial JSON or other error
                    pass
    except Exception as e:
        print(f"[Shim] Failed: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/sandbox/shim.ts
import { connect } from 'bun';

async function main() {
  const socketPath = process.env.BRIDGE_SOCK || './workspace/bridge.sock';
  const command = process.env.SHIM_COMMAND || 'gh';
  const args = process.argv.slice(2);
  const cwd = process.cwd();

  console.error(`[Shim] Connecting to bridge at ${socketPath} for command ${command}...`);

  try {
    const socket = await connect({
      unix: socketPath,
      socket: {
        data(_socket, data) {
          const lines = data.toString().split('\n');
          for (const line of lines) {
            if (!line.trim()) continue;
            try {
              const msg = JSON.parse(line);
              if (msg.type === 'stdout') {
                process.stdout.write(Buffer.from(msg.data, 'base64'));
              } else if (msg.type === 'stderr') {
                process.stderr.write(Buffer.from(msg.data, 'base64'));
              } else if (msg.type === 'exit') {
                process.exit(msg.code);
              } else if (msg.type === 'error') {
                console.error(`[Shim Error] ${msg.message}`);
                process.exit(1);
              }
            } catch {
              // Handle potential partial JSON or multiple messages
            }
          }
        },
        error(_socket, error) {
          console.error(`[Shim] Connection error: ${error}`);
          process.exit(1);
        },
        end(_socket) {
          // Connection closed
        },
      },
    });

    socket.write(
      JSON.stringify({
        command,
        args,
        cwd,
        env: {
          GH_TOKEN: process.env.GH_TOKEN, // In case it's needed, but ideally it's on host
        },
      }),
    );
  } catch (error) {
    console.error(`[Shim] Failed to connect to bridge: ${error}`);
    process.exit(1);
  }
}

main();
